src/main/java/hudson/remoting/RemoteInvocationHandler.java

```diff
@@ -108,15 +133,43 @@
         // if the type is a JDK-defined type, classloader should be for IReadResolve
         if(cl==null || cl==ClassLoader.getSystemClassLoader())
             cl = IReadResolve.class.getClassLoader();
-        return type.cast(Proxy.newProxyInstance(cl, new Class[]{type,IReadResolve.class},
-            new RemoteInvocationHandler(channel,id,userProxy,autoUnexportByCaller,type)));
+        RemoteInvocationHandler handler = new RemoteInvocationHandler(channel, id, userProxy, autoUnexportByCaller, type);
```

Comment:
I couldn't find a case to prove it wasn't. We are dealing with remoting which will involve a network round-trip, so we are not CPU cycle critical and safer is better